# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/platforms/PDMFactory.h
# Commit: 3f08ccb1f141
# Full Hash: 3f08ccb1f141069bfd90474d6023413303307ec4
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2020-11-04 09:55:18
# Regressor Bug: 1595994
# File Overlap Count: 1
# Description:
#   Bug 1674043 - P6. Remove sync PDecoderMananger::Supports API. r=mattwoodrow,bryce,mjf,ipc-reviewers,nika
#   
#   In bug 1595994 we attempted to streamline the ability to determine which decoder was available regardless of the process they would be running in. This was subsequently done via the PDMFactory.
#   
#   As there are several JS API that can query which codec are supported, it requires a synchronous mechanism.
# ==============================================================================

diff -r 33563a16a5c3 -r 3f08ccb1f141 dom/media/platforms/PDMFactory.h
--- a/dom/media/platforms/PDMFactory.h	Mon Nov 02 15:32:59 2020 +0000
+++ b/dom/media/platforms/PDMFactory.h	Wed Nov 04 02:22:33 2020 +0000
@@ -18,6 +18,7 @@
 class PDMFactoryImpl;
 template <class T>
 class StaticAutoPtr;
+enum class RemoteDecodeIn;
 
 class PDMFactory final {
  public:
@@ -25,6 +26,11 @@
 
   PDMFactory();
 
+  // To be called in the content process only, used to determine which PDMs are
+  // usable in their respective process.
+  static already_AddRefed<PDMFactory> PDMFactoryForRdd();
+  static already_AddRefed<PDMFactory> PDMFactoryForGpu();
+
   // Factory method that creates the appropriate PlatformDecoderModule for
   // the platform we're running on. Caller is responsible for deleting this
   // instance. It's expected that there will be multiple
@@ -50,13 +56,42 @@
   static constexpr int kYUV422 = 2;
   static constexpr int kYUV444 = 3;
 
+  /*
+   * All the codecs we support
+   */
+  enum class MediaCodecs {
+    H264,
+    VP9,
+    VP8,
+    AV1,
+    Theora,
+    AAC,
+    MP3,
+    Opus,
+    Vorbis,
+    Flac,
+    Wave,
+
+    SENTINEL,
+  };
+
+  using MediaCodecsSupported = EnumSet<MediaCodecs>;
+
+  static MediaCodecsSupported Supported();
+  static void SetSupported(const MediaCodecsSupported& aSupported);
+
  private:
   virtual ~PDMFactory() = default;
+  // Will set PDM list for the required process.
+  // This is used to determine which PDMs are available on the given process
+  // from the content process.
+  explicit PDMFactory(const RemoteDecodeIn& aProcess);
 
   void CreatePDMs();
   void CreateNullPDM();
   void CreateGpuPDMs();
   void CreateRddPDMs();
+  void CreateContentPDMs();
   void CreateDefaultPDMs();
 
   template <typename DECODER_MODULE, typename... ARGS>
@@ -89,6 +124,14 @@
   static StaticMutex sMonitor;
 };
 
+// Used for IPDL serialization.
+// The 'value' have to be the biggest enum from MediaCodecs.
+template <>
+struct MaxEnumValue<PDMFactory::MediaCodecs> {
+  static constexpr unsigned int value =
+      static_cast<unsigned int>(PDMFactory::MediaCodecs::SENTINEL);
+};
+
 }  // namespace mozilla
 
 #endif /* PDMFactory_h_ */